tests/cases/test_raft.rs

```diff
@@ -3978,3 +3978,153 @@ fn test_remove_learner() {
+// n3 is partitioned, with term 4 and less log, state is candidate
+fn new_prevote_migration_cluster() -> Network {
+    // We intentionally do not enable pre_vote for n3, this is done so in order
+    // to simulate a rolling restart process where it's possible to have a mixed
+    // version cluster with replicas with pre_vote enabled, and replicas without.
+    let mut n1 = new_test_raft_with_prevote(1, vec![1, 2, 3], 10, 1, new_storage(), true);
+    let mut n2 = new_test_raft_with_prevote(2, vec![1, 2, 3], 10, 1, new_storage(), true);
+    let mut n3 = new_test_raft_with_prevote(3, vec![1, 2, 3], 10, 1, new_storage(), false);
+
+    n1.become_follower(1, INVALID_ID);
+    n2.become_follower(1, INVALID_ID);
+    n3.become_follower(1, INVALID_ID);
+
+    let mut nt = Network::new(vec![Some(n1), Some(n2), Some(n3)]);
+
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    // Cause a network partition to isolate n3.
+    nt.isolate(3);
+    nt.send(vec![
+          new_message(1, 1, MessageType::MsgPropose, 1)
+    ]);
+
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+
+    // check state
+    // n1.state == Leader
+    // n2.state == Follower
+    // n3.state == Candidate
+    assert_eq!(nt.peers[&1].state, StateRole::Leader);
+    assert_eq!(nt.peers[&2].state, StateRole::Follower);
+    assert_eq!(nt.peers[&3].state, StateRole::Candidate);
+
+    // check term
+    // n1.Term == 2
+    // n2.Term == 2
+    // n3.Term == 4
+    assert_eq!(nt.peers[&1].term, 2);
+    assert_eq!(nt.peers[&2].term, 2);
+    assert_eq!(nt.peers[&3].term, 4);
+
+    // Enable prevote on n3, then recover the network
+    nt.peers.get_mut(&3).unwrap().pre_vote = true;
+    nt.recover();
+
+    nt
+}
+
+#[test]
+fn test_prevote_migration_can_complete_election() {
+    // n1 is leader with term 2
+    // n2 is follower with term 2
+    // n3 is pre-candidate with term 4, and less log
+    let mut nt = new_prevote_migration_cluster();
+
+    // simulate leader down
+    nt.isolate(1);
+
+    // Call for elections from both n2 and n3.
+    nt.send(vec![new_message(2, 2, MessageType::MsgHup, 0)]);
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+
+    // check state
+    // n2.state == Follower
+    // n3.state == PreCandidate
+    assert_eq!(nt.peers[&2].state,
+               StateRole::Follower, "node 2 state: {:?}, want {:?}",
+               nt.peers[&2].state,
+               StateRole::Follower
+    );
+    assert_eq!(nt.peers[&3].state,
+               StateRole::PreCandidate, "node 3 state: {:?}, want {:?}",
+               nt.peers[&3].state,
+               StateRole::PreCandidate
+    );
+
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+    nt.send(vec![new_message(2, 2, MessageType::MsgHup, 0)]);
+
+    // Do we have a leader?
+    assert_eq!((nt.peers[&2].state, nt.peers[&3].state), (StateRole::Leader, StateRole::Follower));
```

Comment:
I think this is still different from https://github.com/coreos/etcd/pull/8525/files?utf8=%E2%9C%93&diff=unified#diff-ad12a9c9b66e1507ca04edc74b4b1e8aR3386.

Reply:
sorry, I don't know how to fix this since I thought this tuple comparsion might work like `&&`. could you give me some tips?
Will simulating leader down(`nt.solate(1)`) before new election guarantee that?

Reply:
in your test, you require 2 must be the leader, and 3 must be the follower. But in the etcd test, if 2 is the leader, 3 can be not the follower.

Reply:
oh, thank you, I have made a logic mistake here.